A15-binary-logging.md

```diff
@@ -0,0 +1,375 @@
+environment variable or a similarly named flag or option string.
+Turning on binary logging should be a configuration setting and require no
+code changes. The control interface should allow users to change the amount
+and contents of logs.
+
+The logged events correspond to what is visible at the application
+level, and as a result they exclude things like [hedging and
+retries](./A6-client-retries.md).
+
+### Log Format
+
+gRPC logs each RPC in the format defined by proto GrpcLogEntry. Initial/trailing
+metadata and messages are logged separately to match the on-the-wire
+representation.
+
+The proto of GrpcLogEntry:
+
+```proto
+syntax = "proto3";
+
+package grpc.binarylog.v1;
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
+
+option java_multiple_files = true;
+option java_package = "io.grpc.binarylog.v1";
+option java_outer_classname = "BinaryLogProto";
+
+// Log entry we store in binary logs
+message GrpcLogEntry {
+  // Enumerates the type of event
+  // Note the terminology is different from the RPC semantics
+  // definition, but the same meaning is expressed here.
+  enum EventType {
+    EVENT_TYPE_UNKNOWN = 0;
+    // Header sent from client to server
+    EVENT_TYPE_CLIENT_HEADER = 1;
+    // Header sent from server to client
+    EVENT_TYPE_SERVER_HEADER = 2;
+    // Message sent from client to server
+    EVENT_TYPE_CLIENT_MESSAGE = 3;
+    // message sent from server to client
+    EVENT_TYPE_SERVER_MESSAGE = 4;
+    // Signal that client is done sending
+    EVENT_TYPE_CLIENT_HALF_CLOSE = 5;
+    // Trailers sent from server to client.
+    // This marks the end of the RPC. Events may arrive after
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_SERVER_TRAILER = 6;
+    // Signal that the RPC is cancelled.
+    // This marks the end of the RPC. Events may arrive after
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_CANCEL = 7;
+  }
+
+  // Enumerates the entity that generates the log entry
+  enum Logger {
+    LOGGER_UNKNOWN = 0;
+    LOGGER_CLIENT = 1;
+    LOGGER_SERVER = 2;
+  }
+
+  // The timestamp of the binary log message
+  google.protobuf.Timestamp timestamp = 1;
+
+  // Uniquely identifies a call. The value must not be 0 in order to disambiguate
+  // from an unset value.
+  // Each call may have several log entries, they will all have the same call_id.
+  // Nothing is guaranteed about their value other than they are unique across
+  // different RPCs in the same gRPC process.
+  int64 call_id = 2;
+
+  // The entry sequence id for this call. The first GrpcLogEntry has a
+  // value of 1, to disambiguate from an unset value. The purpose of
+  // this field is to detect missing entries in environments where
+  // durability or ordering is not guaranteed.
+  uint32 sequence_id_within_call = 3;
```

Comment:
Can we make this also 64 bit so the same ID generator can be reused?

Reply:
int64 makes sense. Done.

Reply:
It's a bit of a nit, but should both of these be uint64 instead? I don't think we want to use negative numbers for either, and it's best to make things unexpressable if you don't want them.

Reply:
Agreed, a negative call_id value is probably just asking for trouble. I'll make it a uint64.